bpo-36004: Add date.fromisocalendar

pganssle opened the pull request:
This commit implements the first version of `date.fromisocalendar`, the inverse function for `date.isocalendar`. It is currently missing error checking for the case of of invalid ISO dates in week 53.

To Do:
- [x]  Validate ISO datetimes
- [x] Test for `TypeError`
- [x] Add documentation
- [x] Add news entry

~~Other than these known errors, the existing code can be reviewed.~~ Ready to go.

[bpo-36004](https://bugs.python.org/issue36004)


<!-- issue-number: [bpo-36004](https://bugs.python.org/issue36004) -->
https://bugs.python.org/issue36004
<!-- /issue-number -->

pganssle committed:
Add date.fromisocalendar and tests

This commit implements the first version of date.fromisocalendar, the
inverse function for date.isocalendar. It is currently missing error
checking for the case of of invalid iso dates in week 53.

bpo-36004: https://bugs.python.org/issue36004

pablogsal commented on Modules/_datetimemodule.c:
```diff
@@ -845,6 +845,17 @@ parse_isoformat_time(const char *dtstr, size_t dtlen, int *hour, int *minute,
+
+    int day_offset = (*month_week - 1)*7 + *day - 1;
+
+    ord_to_ymd(day_1 + day_offset, year, month_week, day);
```
This will overflow (and core dump) if the year is very big. Example:

```python
>>> import datetime
>>> datetime.date.fromisocalendar(9999999,1,1)
python: /home/pablogsal/github/cpython/Modules/_datetimemodule.c:307: ord_to_ymd: Assertion `ordinal >= 1' failed.
[1]    19212 abort (core dumped)  ./python
```

pganssle replied:
Hm, I was under the impression that `i` arguments to the `PyArg_Parse*` macros were checked for overflow, maybe I got that wrong?

The only reason I didn't have a test for this yet was that I hadn't decided whether it was fine to surface an `OverflowError` or whether I should catch and re-raise `ValueError`, since any overflow is a domain error anyway. I think catch and re-raise is the right way to go.

I'll update the PR tomorrow.

pganssle replied:
Hm, I can't reproduce the core dump, but actually I realize what's happening here, it's not that any of the inputs is overflowing, it's that the number of ordinal days that is overflowing. That's easy enough to fix by more aggressively checking the bounds on the input. Good catch.

pablogsal replied:
The inputs themselves cannot overflow, if you try to pass a bigger value of the year you will receive different error messages telling that a C `int` cannot be constructed. Is the moment you are adding two valid C `ints` (`day_1 + day_offset`) or when you call `iso_week1_monday` the moment it overflows.

The core dump happens because of the assert in `_datetimemodule.c:307` fails because the overflow has converted the number into a negative one. You need to compile in debug mode to see it.

pganssle replied:
@pablogsal Yeah, I think the general gist of what you are saying is correct, but I think it's actually overflowing when the year gets converted to ordinal date, not the addition (which can add a maximum of ~400 days because of the bounds of the inputs). Either way, it's fixed now.

pganssle committed:
Add early year check to fromisocalendar

This avoids an overflow error in ordinal calculations in the C
implementation.

pganssle committed:
Covert OverflowError into ValueError

All overflow errors in this context are also value errors, so it is
preferable to be uniform in the type of error raised.

pganssle committed:
Add bounds checking for ISO long vs. short weeks

pganssle commented on Lib/datetime.py:
```diff
@@ -2115,6 +2138,10 @@ def __reduce__(self):
 
 
+def _iso_long_year_helper(year):
+    return (year + (year // 4) - (year // 100) + (year // 400)) % 7
```
TODO: Definitely need a comment explaining what this function does and means. Possibly define it *in* `fromisocalendar` instead of here.

ZackerySpytz commented on Modules/_datetimemodule.c:
```diff
@@ -2986,6 +2986,61 @@ date_fromisoformat(PyObject *cls, PyObject *dtstr)
 }
 
+
+static int _iso_long_year_helper(int year) {
```
For functions, the opening brace needs to be on its own line. See [PEP 7](https://www.python.org/dev/peps/pep-0007/).

ZackerySpytz replied:
(The return type also needs to be on its own line.)

pganssle replied:
Thanks for the review, good catch. I ended up removing this function in the end so it is moot now, but I did move the `{` to its own line in the `date_fromisocalendar`.

ZackerySpytz commented on Modules/_datetimemodule.c:
```diff
@@ -2986,6 +2986,61 @@ date_fromisoformat(PyObject *cls, PyObject *dtstr)
+    }
+
+    if (day <= 0 || day >= 8) {
+        PyErr_Format(PyExc_ValueError, "Invalid day: %d (range is [1, 7])");
```
It looks like you're missing an argument.

pganssle replied:
Oof. Now I'm tempted to add a test for this, even though I am generally against the idea of testing that error messages say a specific thing.

I think I'll just fix it and leave it untested. Good catch!

pganssle committed:
Add more comments with test cases

pganssle committed:
Switch to new method of detecting ISO long years

This is equivalent but uses only existing helper functions and in many
cases will be slightly more efficient.

pganssle committed:
Fix missing argument in PyErr_Format

pganssle committed:
Add test for TypeError handling

pganssle committed:
Add documentation for date(time).fromisocalendar

pganssle committed:
Add NEWS entry

pganssle commented on Lib/test/datetimetester.py:
```diff
@@ -1795,6 +1795,75 @@ def test_fromisoformat_fails_typeerror(self):
+            (None, 1, 1),
+            (2019, None, 1),
+            (2019, 1, None),
+            (2019.0, 1, 1),
```
One thing I'll note is that this is consistent with `datetime.date` itself not accepting floats. I'm not sure how much I agree with that, I just wanted to put it in the tests to ensure consistency between the C and Python versions.

pablogsal reviewed (commented):
I think we are missing an entry in `Doc/whatsnew/3.8.rst`

pganssle committed:
Add What's New entry for fromisocalendar

vstinner commented on Lib/datetime.py:
```diff
@@ -884,6 +884,40 @@ def fromisoformat(cls, date_string):
+
+        This is the inverse of the date.isocalendar() function"""
+        # Year is bounded this way because 9999-12-31 is (9999, 52, 5)
+        if not 0 < year < 10000:
```
MINYEAR/MAXYEAR should be used here, no?

pganssle replied:
Yes, good point.

vstinner commented on Lib/test/datetimetester.py:
```diff
@@ -1795,6 +1795,75 @@ def test_fromisoformat_fails_typeerror(self):
+            (2004, 1, 1),       # Leap year
+            (2004, 12, 31),
+            (1, 1, 1),
+            (9999, 12, 31),
```
Maybe use MINYEAR/MAXYEAR here.

pganssle replied:
Instead of converting these over, I have just added a `MINYEAR` and `MAXYEAR` test. Even though it's redundant, I like to have the explicit callout of the current boundaries, so that if `MINYEAR` or `MAXYEAR` get modified in a way that breaks backwards compatibility, it will raise an error.

vstinner commented on Lib/test/datetimetester.py:
```diff
@@ -1795,6 +1795,75 @@ def test_fromisoformat_fails_typeerror(self):
+            (2019.0, 1, 1),
+            (2019, 1.0, 1),
+            (2019, 1, 1.0),
+        ]
```
Would it be possible to use 3 loops to test all combinations, rather than generate these combinations manualy?

pganssle replied:
OK, I've converted it over. I think it's a little harder to understand what's going on in the version where the test cases are generated (compared to manual), but using a loop also has its advantages.

pganssle committed:
Use MINYEAR and MAXYEAR in bounds checks

pganssle committed:
Add tests for MINYEAR and MAXYEAR

pganssle committed:
Use generated test cases instead of manual list

vstinner reviewed (approved):
LGTM.

Side note: @pganssle: maybe it would be interesting to convert datetime to Argument Clinic to provide better docstrings.
